### ISSUE TITLE
Create injection of Func<IService>

### ISSUE THREAD

**Issue by Alex2357** [moq/Moq.AutoMocker#21]
This is a useful feature. This lines could be avoided if it would be able to inject Func<>

```
        var sender = CreateStub<IEmailSender>();
        Use<Func<IEmailSender>>(() => sender);
        base.SutInitialize();
```

What if you have got 3 or more dependencies like this?

These are methods of my base class

```
    protected T CreateStub<T>()
        where T : class
    {
        return new Mock<T>().Object;
    }
    protected virtual void SutInitialize()
    {
        Sut = CreatePartial<TSut>();
    }

    protected T CreatePartial<T>()
        where T : class
    {
        var instance = Mocker.CreateSelfMock<T>();
        var mock = Mock.Get(instance);
        //I need to add it as it can be I need to do setup against that partial.
        Mocker.Use(mock);
        mock.CallBase = true;
        return instance;
    }

    protected void Use<T>(T service)
        where T : class
    {
        Mocker.Use(service);
    }
```

Kind regards,
Alex.

**Comment by adamhewitt627** [comment]
@Alex2357 I started some work on this, as some of my projects would benefit from it. As I understand, you would like something like:

```csharp
public class SystemUnderTest
{
    public SystemUnderTest(Func<IService> func)
    {
        //Instead of null/failure, func is automatically resolved to (effectively)
        //
        //      () => mocker.Get<IService>()
    }
}

void MyUnitTest()
{
    //Arrange
    AutoMocker mocker = new AutoMocker();

    var sut = mocker.CreateInstance<SystemUnderTest>(); //automatically injects Func<IService>

    //Act & assert
}
```

**Comment by Alex2357** [comment]
Yes, that looks exactly what I wanted. I just did not touch automocker for 1 year.

**Comment by adamhewitt627** [comment]
Well, welcome back. @Keboo and I have recently taken on maintenance of this and hope to build on a lot of what is already here.

For my own project, injecting `Func` ran into some difficulty with determining whether to call `Get` or `CreateInstance`. Currently, `() => typeof(T).IsClass ? mocker.CreateInstance<T>() : mocker.Get<T>()` works but that is really limited to my scope/requirements. In addition:

1. Should this be enabled by default?
2. If configurable, what should that API look like?
3. Can we (somehow?) include the `Func` in a setup in order to assert on calls?
4. Should `Lazy<T>` resolve with the same rules?


On my own, I chose to add this using extension methods to mimic a fluent API for more fine-grained control of mocks per-test.

```csharp
public static AutoMocker WithFunc<TResult>(this AutoMocker mocker, Func<TResult> func = null) where TResult : class
{
    //Could/should probably be a more universal creation, handle failures, etc.
    mocker.Use(func ?? (() => typeof(TResult).IsClass ? mocker.CreateInstance<TResult>() : mocker.Get<TResult>()));
    return mocker;
}

void Test()
{
    var mocker = new AutoMocker()
        .WithFunc<IService>()  //Uses default creation as shown above
        .WithLazy<IService2>(() => new Service2());
}
```

Not as pretty as the request, but adds _some_ elegance and simplicity without complicating the API.

**Comment by Alex2357** [comment]
It seems you did not the same thing you posted in the first your post. If I remember correctly, I wanted it to be automatically done for Func<ISomeInterface>.
You wanna add a method what is not automatic. Just think you may add a new Func<I...> dependency in the future and you have to update all the tests.
I really like automocker for ability not to specify all the dependencies. Only for the cases when something needs to be tweaked I had to do some coding.

**Comment by adamhewitt627** [comment]
Yes, sorry for being unclear, but that second post was meant to:

1. Illustrate the challenge in defining what the contents of the injected `Func` should be. (and raise questions)
2. Show how I've done it in the past to limit code repetition and maintain flexibility.

**Comment by adamhewitt627** [comment]
I spent some time discussing this with @Keboo, and I think we have a direction. I've started on the implementation with: https://github.com/moq/Moq.AutoMocker/compare/mock-resolvers

In short, AutoMocker adds `IMockResolver` to allow extending to include additional resolutions as needed. Some would be defined by AutoMocker (`Func<T>`, `Lazy<T>` to begin with) with *default behavior* defined by whatever the initial set of resolvers is.

```csharp
// Default Resolvers
var mocker = new AutoMocker();

// Custom ResolverSet
mocker = new AutoMocker
{
    Resolvers = {
        // Func & Lazy might be included by default, I'm undecided on how I feel about that
        new FuncResolver(),
        new LazyResolver(),
        new MyCustomResolver(), //if you need to define your own
    }
};
```

**Comment by adamhewitt627** [comment]
This is fixed with #48.
